src/data/finset.lean

```diff
@@ -586,6 +622,51 @@ def attach (s : finset α) : finset {x // x ∈ s} := ⟨attach s.1, nodup_attac
 
 @[simp] theorem attach_empty : attach (∅ : finset α) = ∅ := rfl
 
+/-! ### piecewise -/
+section piecewise
+
+/-- `s.piecewise f g` is the function equal to `f` on the finset `s`, and to `g` on its complement. -/
+def piecewise {α : Type*} {δ : α → Sort*} (s : finset α) (f g : Πi, δ i) [∀j, decidable (j ∈ s)] :
+  Πi, δ i :=
+λi, if i ∈ s then f i else g i
+
+variables {δ : α → Sort*} (s : finset α) (f g : Πi, δ i)
+
+@[simp] lemma piecewise_insert_self [decidable_eq α] {j : α} [∀i, decidable (i ∈ insert j s)] :
+  (insert j s).piecewise f g j = f j :=
+by simp [piecewise]
+
+@[simp] lemma piecewise_empty [∀i : α, decidable (i ∈ (∅ : finset α))] : piecewise ∅ f g = g :=
+by { ext i, simp [piecewise] }
+
+variable [∀j, decidable (j ∈ s)]
+
+@[elim_cast] lemma piecewise_coe [∀j, decidable (j ∈ (↑s : set α))] :
+  (↑s : set α).piecewise f g = s.piecewise f g :=
+by { ext, congr }
+
+@[simp] lemma piecewise_eq_of_mem {i : α} (hi : i ∈ s) : s.piecewise f g i = f i :=
+by simp [piecewise, hi]
+
+@[simp] lemma piecewise_eq_of_not_mem {i : α} (hi : i ∉ s) : s.piecewise f g i = g i :=
+by simp [piecewise, hi]
+
+@[simp] lemma piecewise_insert_of_ne [decidable_eq α] {i j : α} [∀i, decidable (i ∈ insert j s)]
+  (h : i ≠ j) : (insert j s).piecewise f g i = s.piecewise f g i :=
+by { simp [piecewise, h], congr }
+
+lemma piecewise_insert [decidable_eq α] (j : α) [∀i, decidable (i ∈ insert j s)] :
+  (insert j s).piecewise f g = function.update (s.piecewise f g) j (f j) :=
```

Comment:
I wonder if it's hard to reuse `set.piecewise_insert` here, and similarly for `powerset_insert`.

Reply:
I changed to use the set version to deduce the finset version by coercion here. I also tried in `powerset_insert`, but it didn't work out well because there is a finset of finsets on one side, and a set of sets on the other side, so the translation is not direct at all.